File: mqtt-logger/logger.py
#!/usr/bin/env python3
"""Subscribes to all MQTT topics and persists every message to PostgreSQL."""

import logging
import os
import sys
import time

import psycopg2
import paho.mqtt.client as mqtt

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(message)s",
    stream=sys.stdout,
)
log = logging.getLogger(__name__)

MQTT_HOST = os.environ.get("MQTT_HOST", "mqtt")
MQTT_PORT = int(os.environ.get("MQTT_PORT", "1883"))
MQTT_USER = os.environ["MQTT_USER"]
MQTT_PASSWORD = os.environ["MQTT_PASSWORD"]
MQTT_TOPIC = os.environ.get("MQTT_TOPIC", "#")

PGHOST = os.environ.get("PGHOST", "postgres-1")
PGPORT = int(os.environ.get("PGPORT", "5432"))
PGUSER = os.environ["PGUSER"]
PGPASSWORD = os.environ["PGPASSWORD"]
PGDATABASE = os.environ.get("PGDATABASE", "mqtt_store")

CREATE_TABLE_SQL = """
CREATE TABLE IF NOT EXISTS mqtt_messages (
    id          BIGSERIAL    PRIMARY KEY,
    topic       TEXT         NOT NULL,
    payload     BYTEA,
    qos         SMALLINT     NOT NULL,
    retain      BOOLEAN      NOT NULL,
    received_at TIMESTAMPTZ  NOT NULL DEFAULT NOW()
);
"""

INSERT_SQL = """
INSERT INTO mqtt_messages (topic, payload, qos, retain)
VALUES (%s, %s, %s, %s)
"""


def connect_pg():
    while True:
        try:
            conn = psycopg2.connect(
                host=PGHOST,
                port=PGPORT,
                user=PGUSER,
                password=PGPASSWORD,
                dbname=PGDATABASE,
            )
            conn.autocommit = True
            with conn.cursor() as cur:
                cur.execute(CREATE_TABLE_SQL)
            log.info("Connected to PostgreSQL (%s/%s), table ready.", PGHOST, PGDATABASE)
            return conn
        except Exception as exc:
            log.warning("PostgreSQL not ready (%s), retrying in 5s…", exc)
            time.sleep(5)


pg_conn = None


def get_cursor():
    global pg_conn
    if pg_conn is None or pg_conn.closed:
        pg_conn = connect_pg()
    return pg_conn.cursor()


def on_connect(client, userdata, flags, reason_code, properties=None):
    if reason_code == 0:
        log.info("Connected to MQTT broker %s:%s, subscribing to '%s'", MQTT_HOST, MQTT_PORT, MQTT_TOPIC)
        client.subscribe(MQTT_TOPIC, qos=1)
    else:
        log.error("MQTT connection refused, reason code: %s", reason_code)


def on_message(client, userdata, msg):
    global pg_conn
    try:
        cur = get_cursor()
        cur.execute(INSERT_SQL, (msg.topic, msg.payload, msg.qos, msg.retain))
        cur.close()
        log.debug("Stored [%s] %d bytes", msg.topic, len(msg.payload or b""))
    except Exception as exc:
        log.error("Failed to store message from '%s': %s", msg.topic, exc)
        pg_conn = None


def main():
    global pg_conn
    pg_conn = connect_pg()

    client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION2)
    client.username_pw_set(MQTT_USER, MQTT_PASSWORD)
    client.on_connect = on_connect
    client.on_message = on_message

    while True:
        try:
            client.connect(MQTT_HOST, MQTT_PORT, keepalive=60)
            client.loop_forever()
        except Exception as exc:
            log.warning("MQTT error (%s), reconnecting in 5s…", exc)
            time.sleep(5)


if __name__ == "__main__":
    main()

File: mqtt-logger/logger.ts
import mqtt, { MqttClient } from "mqtt";
import { Pool } from "pg";

function requireEnv(name: string): string {
  const val = process.env[name];
  if (!val) {
    console.error(`Missing required environment variable: ${name}`);
    process.exit(1);
  }
  return val;
}

const MQTT_HOST     = process.env.MQTT_HOST  ?? "mqtt";
const MQTT_PORT     = parseInt(process.env.MQTT_PORT ?? "1883", 10);
const MQTT_USER     = requireEnv("MQTT_USER");
const MQTT_PASSWORD = requireEnv("MQTT_PASSWORD");
const MQTT_TOPIC    = process.env.MQTT_TOPIC ?? "#";

const pool = new Pool({
  host:     process.env.PGHOST     ?? "postgres-1",
  port:     parseInt(process.env.PGPORT ?? "5432", 10),
  user:     requireEnv("PGUSER"),
  password: requireEnv("PGPASSWORD"),
  database: process.env.PGDATABASE ?? "mqtt_store",
});

const CREATE_TABLE_SQL = `
  CREATE TABLE IF NOT EXISTS mqtt_messages (
    id          BIGSERIAL   PRIMARY KEY,
    topic       TEXT        NOT NULL,
    payload     BYTEA,
    qos         SMALLINT    NOT NULL,
    retain      BOOLEAN     NOT NULL,
    received_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
  );
`;

async function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function initDb(): Promise<void> {
  while (true) {
    try {
      await pool.query(CREATE_TABLE_SQL);
      console.log(`[pg] Connected to ${pool.options.host}/${pool.options.database}, table ready.`);
      return;
    } catch (err) {
      console.warn(`[pg] Not ready (${(err as Error).message}), retrying in 5s…`);
      await sleep(5000);
    }
  }
}

function startMqtt(): void {
  const client: MqttClient = mqtt.connect(`mqtt://${MQTT_HOST}:${MQTT_PORT}`, {
    username: MQTT_USER,
    password: MQTT_PASSWORD,
    reconnectPeriod: 5000,
  });

  client.on("connect", () => {
    console.log(`[mqtt] Connected to ${MQTT_HOST}:${MQTT_PORT}, subscribing to '${MQTT_TOPIC}'`);
    client.subscribe(MQTT_TOPIC, { qos: 1 }, (err) => {
      if (err) console.error("[mqtt] Subscribe error:", err.message);
    });
  });

  client.on("message", (topic, payload, packet) => {
    pool
      .query(
        "INSERT INTO mqtt_messages (topic, payload, qos, retain) VALUES ($1, $2, $3, $4)",
        [topic, payload, packet.qos, packet.retain],
      )
      .catch((err: Error) =>
        console.error(`[mqtt] Failed to store '${topic}': ${err.message}`),
      );
  });

  client.on("error",     (err) => console.error("[mqtt] Error:", err.message));
  client.on("reconnect", ()    => console.log("[mqtt] Reconnecting…"));
}

async function main(): Promise<void> {
  await initDb();
  startMqtt();
}

main();
